src/openrct2-ui/windows/MazeConstruction.cpp

```diff
@@ -83,379 +83,351 @@ static rct_widget window_maze_construction_widgets[] = {
-    gMapSelectFlags &= ~MAP_SELECT_FLAG_ENABLE_ARROW;
+        gMapSelectFlags &= ~MAP_SELECT_FLAG_ENABLE;
+        gMapSelectFlags &= ~MAP_SELECT_FLAG_ENABLE_ARROW;
 
-    CoordsXYZD entranceOrExitCoords = ride_get_entrance_or_exit_position_from_screen_position(screenCoords);
-    if (entranceOrExitCoords.IsNull())
-        return;
+        CoordsXYZD entranceOrExitCoords = ride_get_entrance_or_exit_position_from_screen_position(screenCoords);
+        if (entranceOrExitCoords.IsNull())
+            return;
 
-    if (gRideEntranceExitPlaceDirection == INVALID_DIRECTION)
-        return;
+        if (gRideEntranceExitPlaceDirection == INVALID_DIRECTION)
+            return;
 
-    ride_id_t rideIndex = gRideEntranceExitPlaceRideIndex;
+        ride_id_t rideIndex = gRideEntranceExitPlaceRideIndex;
 
-    auto rideEntranceExitPlaceAction = RideEntranceExitPlaceAction(
-        entranceOrExitCoords, direction_reverse(entranceOrExitCoords.direction), rideIndex, gRideEntranceExitPlaceStationIndex,
-        gRideEntranceExitPlaceType == ENTRANCE_TYPE_RIDE_EXIT);
+        auto rideEntranceExitPlaceAction = RideEntranceExitPlaceAction(
+            entranceOrExitCoords, direction_reverse(entranceOrExitCoords.direction), rideIndex,
+            gRideEntranceExitPlaceStationIndex, gRideEntranceExitPlaceType == ENTRANCE_TYPE_RIDE_EXIT);
 
-    rideEntranceExitPlaceAction.SetCallback([=](const GameAction* ga, const GameActions::Result* result) {
-        if (result->Error != GameActions::Status::Ok)
-            return;
+        rideEntranceExitPlaceAction.SetCallback([=](const GameAction* ga, const GameActions::Result* result) {
+            if (result->Error != GameActions::Status::Ok)
+                return;
+
+            OpenRCT2::Audio::Play3D(OpenRCT2::Audio::SoundId::PlaceItem, result->Position);
+
+            auto currentRide = get_ride(rideIndex);
+            if (currentRide != nullptr && ride_are_all_possible_entrances_and_exits_built(currentRide))
+            {
+                tool_cancel();
+                if (currentRide->GetRideTypeDescriptor().HasFlag(RIDE_TYPE_FLAG_HAS_NO_TRACK))
+                    window_close_by_class(WC_RIDE_CONSTRUCTION);
+            }
+            else
+            {
+                gRideEntranceExitPlaceType = gRideEntranceExitPlaceType ^ 1;
+                window_invalidate_by_class(WC_RIDE_CONSTRUCTION);
+                gCurrentToolWidget.widget_index = (gRideEntranceExitPlaceType == ENTRANCE_TYPE_RIDE_ENTRANCE)
+                    ? WIDX_MAZE_ENTRANCE
+                    : WIDX_MAZE_EXIT;
+
+                WindowMazeConstructionUpdatePressedWidgets();
+            }
+        });
+        auto res = GameActions::Execute(&rideEntranceExitPlaceAction);
+    }
 
-        OpenRCT2::Audio::Play3D(OpenRCT2::Audio::SoundId::PlaceItem, result->Position);
+    void WindowMazeConstructionConstruct(int32_t direction)
+    {
+        int32_t x, y, z, flags_, mode;
```

Comment:
I had to rename flags because it shadows a member. I don't really like flags_ but didn't have a better idea, so feel free to suggest something else.

Reply:
You can rename it to actionFlags.

Reply:
Done.